src/renderer/base/FontCache.h

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+#include <dwrite_3.h>
+
+namespace Microsoft::Console::Render::FontCache
+{
+    namespace details
+    {
+        struct Cache
+        {
+            void Invalidate() noexcept
+            {
+                _lock.lock();
+                _fontCollection.reset();
+                _lock.unlock();
+            }
+
+            wil::com_ptr<IDWriteFontCollection> Get(bool forceUpdate)
+            {
+                std::lock_guard guard{ _lock };
+
+                if (!_fontCollection || forceUpdate)
+                {
+                    _init(forceUpdate);
+                    FAIL_FAST_IF(!_fontCollection);
+                }
+
+                return _fontCollection;
+            }
+
+        private:
+            void _init(bool forceUpdate)
+            {
+                // DWRITE_FACTORY_TYPE_SHARED _should_ return the same instance every time.
+                wil::com_ptr<IDWriteFactory> factory;
+                THROW_IF_FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(factory), reinterpret_cast<::IUnknown**>(factory.addressof())));
+
+                wil::com_ptr<IDWriteFontCollection> systemFontCollection;
+                THROW_IF_FAILED(factory->GetSystemFontCollection(systemFontCollection.addressof(), forceUpdate));
+
+                // IDWriteFactory5 is supported since Windows 10, build 15021.
+                const auto factory5 = factory.try_query<IDWriteFactory5>();
+                if (!factory5)
+                {
+                    _fontCollection = systemFontCollection;
+                    return;
+                }
+
+                // Cache nearby files.
+                if (!_nearbyFilesFound)
+                {
+                    _nearbyFilesFound = true;
+
+                    const std::filesystem::path module{ wil::GetModuleFileNameW<std::wstring>(nullptr) };
+                    const auto folder{ module.parent_path() };
+
+                    for (const auto& p : std::filesystem::directory_iterator(folder))
+                    {
+                        if (til::ends_with(p.path().native(), L".ttf"))
+                        {
+                            wil::com_ptr<IDWriteFontFile> fontFile;
+                            if (SUCCEEDED_LOG(factory5->CreateFontFileReference(p.path().c_str(), nullptr, fontFile.addressof())))
```

Comment:
will this log generate a telemetry event using WIL's registered logging handler? Should we log on success? Am I getting this backwards, and it logs if failed only?

Reply:
Yeah, `SUCCEEDED_LOG` only logs on failure. It's a shorthand for `SUCCEEDED(LOG_IF_FAILED(hr))`.